### DISCUSSION TITLE
LocalizationProvider - Bring your own date management library

### DISCUSSION THREAD

**Discussion by devrnt** [mantinedev/mantine#333] [Feedback]
## Feature Request
Bring your own date management library:  It's not a big deal to have multiple date libraries, I've read https://github.com/mantinedev/mantine/discussions/198, but it becomes more "pleasant" and flexible to let the developer choose their own preferred date library. This can also potentially result in appealing to a wider audience.

Let's introduce a `LocalizationProvider` in mantine.dev!

## Plan of Action
Shipping this feature would result in a breaking change, however, we are able to foresee a fallback (to the current `dayjs` library). 

I'm heavily inspired by the approach of [https://mui.com/](mui) (no need to reinvent the wheel here).

1. Implement the `LocationProvider` which takes in a preffered date library adapter as prop, see adapters: https://github.com/dmtrKovalenko/date-io.
2. Migrate all the components that are dependent on the current date library (dayjs). Would mainly be the `@mantine/dates` components. These components will make use of the adapter api
3. Document everything

I've been doing some work already and I plan to continue working on this feature on condition that you green light.

Also awesome package, kudos!

**Comment by rtivital** [comment]
Hi @devrnt, thanks for proposal! I have not used date-io library, so can you please clarify how localization process would work with different libraries (let's take date-fns and dayjs as examples) – I'm most interested in locales tree shaking part since it is not obvious from date-io docs.

**Reply by devrnt** [reply]
Hi, `@date/io` is an extra abstraction layer over various date management libraries (see https://github.com/dmtrKovalenko/date-io#projects). 

So in mantine we can use the uniform `@date-io` api regardless of 
the chosen date library (adapter). We would install and export all the "supported" adapter in mantine itself, however I quote 
>Yes, you will install all of them as dependencies. But every adapter is 50kb unpacked npm module and relieas to the library as for optional peer dependency. It won't be included in user bundle until user will choose which library he want's to use.

And see `package.json` https://github.com/dmtrKovalenko/date-io/tree/master/packages/date-fns

##### Example
```jsx
// App.tsx

import { AdapterDateFns } from "@mantine/dates/adapters";

const dateFnsAdapter = new DateFnsAdapter({ locale: "nl" });

<LocalizationProvider dateAdapter={dateFnsAdapter}>
    // ...
</LocalizationProvider />
```

```jsx
// DateRangePicker.tsx

// This context value contains the localization util
import { useLocalizationAdapter } from '../LocalizationProvider/LocalizationProvider';

const DateRangePicker = () => {
  // See api of utils: https://github.com/dmtrKovalenko/date-io#interface
  const { utils } = useLocalizationAdapter();

  return (
     // ...
  ) 
}
 
```

**Reply by rtivital** [reply]
Thanks for explanation. I've researched this feature and came to the conclusion that it will:
- require additional development for your build system
- introduce new dependencies to the project while keeping old dayjs dependency (we need default date library)
- require more setup for localization
- be harder to maintain 

So I would prefer to leave things as is. But thanks for proposal!

**Reply by vorant94** [reply]
does the current status of tc39 proposal for temporal ([link](https://github.com/tc39/proposal-temporal)) change anything on your decision here (maybe in the future when temporal becomes widely adopted standard)?

having an option to just drop all of date utility libs from the bundle sounds pretty intriguing to me even though i agree these libraries usually has very little footprint

**Reply by rtivital** [reply]
It is a proposal, not working yet, years before any significant browser support.

**Reply by vorant94** [reply]
yeah, i'm aware that it is still a proposal, but it is already stage 3 proposal, which means that the design is finished and most likely it is just a matter of time for becoming widely adopted by browsers

so in general my question more about if someone implements `LocalizationProvider` now or in the future before/after temporal becomes a standard, will it have a chance of merging into `@mantine/dates` or all the drawbacks you listed above still outweigh the benefit of such feature from your point of view?

**Reply by rtivital** [reply]
It is years before significant browser support, what's the point of discussing this now? Would it be any use for you if I say "Oh, yes, it is so cool, we would do that in 5 years"?

**Reply by vorant94** [reply]
since mantine is OSS anyone can decide to spend his/her spare time to implement feature they think will benefit the library. but every such change comes with a number of burdens:

- burden on contributor to implement it
- burden on maintainer to review and support it
- burden on other lib users to adapt it (if there are changes in public API or whatever)

while topic-starter or me or anybody else who wants this feature to be there can take the first burden on, the arguments you listed above appeal more to the second and third burdens. so even if feature is implemented without maintainers effort the PR can be declined due to those arguments and because of this risk it is questionable to even begin thinking about implementation

so the point of discussing it now is to just clarify whether this feature (if implemented by some volunteer contributor) has chance to be merged in the light of this proposal despite all the trade-offs you listed either now before temporal is widely adopted or when temporal becomes widely adopted. or it won't be merged even after temporal adoption because of backward compatibility or any other additional valid reasons

unless it is clarified nobody for sure won't take to work on it, once its clarified there is a possibility of someone taking it

**Reply by rtivital** [reply]
This is not something that I can accept as a contribution – such feature requires library rewrite and in-depth understanding of the source code. Once Temporal is widely available, it will be considered for a migration.
